Signal/src/ViewControllers/NewContactThreadViewController.m

```diff
@@ -295,46 +320,166 @@ - (void)updateTableContents
+        }
+    } else {
+        for (OWSTableSection *section in [self collatedContactsSections]) {
+            [contents addSection:section];
+        }
+        contents.sectionForSectionIndexTitleBlock = ^NSInteger(NSString *_Nonnull title, NSInteger index) {
+            // We have at least one non-collated section - the "Find By Phone Number / Inbite Friends" section
+            // If we don't account for these, our collation indices will be offset.
+
+            // Count the sections which are not part of our collation.
+            NSInteger extraSections
+                = (NSInteger)contents.sections.count - (NSInteger)self.collation.sectionTitles.count;
+
+            NSInteger sectionIndex = [self.collation sectionForSectionIndexTitleAtIndex:index] + extraSections;
+            if (sectionIndex < 0) {
+                // Sentinal in case we change our section ordering in a surprising way.
+                OWSFail(@"Unexpected negative section index");
+                return 0;
+            }
+            if (sectionIndex >= (NSInteger)contents.sections.count) {
+                // Sentinal in case we change our section ordering in a surprising way.
+                OWSFail(@"Unexpectedly large index");
+                return 0;
+            }
 
+            return sectionIndex;
+        };
+        contents.sectionIndexTitlesForTableViewBlock = ^NSArray<NSString *> *_Nonnull
+        {
+            return self.collation.sectionTitles;
+        };
+    }
+
+    self.tableViewController.contents = contents;
+}
+
+- (NSArray<OWSTableSection *> *)collatedContactsSections
+{
+    if (self.contactsViewHelper.signalAccounts.count < 1) {
+        // No Contacts
+        OWSTableSection *contactsSection = [OWSTableSection new];
+
+        if (self.contactsViewHelper.contactsManager.isSystemContactsAuthorized
+            && self.contactsViewHelper.hasUpdatedContactsAtLeastOnce) {
+            
+            [contactsSection
+             addItem:[OWSTableItem
+                      softCenterLabelItemWithText:NSLocalizedString(@"SETTINGS_BLOCK_LIST_NO_CONTACTS",
+                                                                    @"A label that indicates the user has no Signal contacts.")
+                      customRowHeight:self.actionCellHeight]];
+        }
+        
+        return @[ contactsSection ];
+    }
+    __weak NewContactThreadViewController *weakSelf = self;
+    
+    NSMutableArray<OWSTableSection *> *contactSections = [NSMutableArray new];
+    
+    NSMutableArray<NSMutableArray<SignalAccount *> *> *collatedSignalAccounts = [NSMutableArray new];
+    for (NSUInteger i = 0; i < self.collation.sectionTitles.count; i++) {
+        collatedSignalAccounts[i] = [NSMutableArray new];
+    }
+    for (SignalAccount *signalAccount in self.contactsViewHelper.signalAccounts) {
+        NSInteger section =
+            [self.collation sectionForObject:signalAccount collationStringSelector:@selector(stringForCollation)];
```

Comment:
Let's make sure the test plan includes contacts...

* without a first or last name.
* with a display name that starts with whitespace.
* with a display name that starts with a chinese high-code-plane character.
* with a display name that starts with punctuation.
* with a display name that starts with emoji.

Reply:
:+1: